### PR TITLE
[GDI32_APITEST] Add CursorIcon GetObject test case

### PR DESCRIPTION
**Pull request by zecarlos1957**
[GDI32_APITEST] Add CursorIcon GetObject test case
CORE-17752

### CODE REVIEW

**Comment by GeoB99** [comment]
Could you please post some screenshots of the respective tests running in Windows Server 2003 and ReactOS?

File: modules/rostests/apitests/gdi32/GetObject.c [hunk 1]
```diff
@@ -701,11 +701,23 @@ Test_Region(void)
     DeleteObject(hRgn);
 }
 
+void Test_CursorIcon()
+{
+    ICONINFO info;
+    BITMAP bmp;
+    HCURSOR hCursor = LoadImage(NULL, MAKEINTRESOURCE(OCR_NORMAL), IMAGE_CURSOR, 0, 0, LR_DEFAULTSIZE|LR_SHARED);
```
Comment (GeoB99) [inline modules/rostests/apitests/gdi32/GetObject.c:708]: You're currently using the Wide (Unicode) variant of `GetObject`, so please use the Unicode variant for this one as well.

File: modules/rostests/apitests/gdi32/GetObject.c [hunk 2]
```diff
@@ -716,5 +728,6 @@ START_TEST(GetObject)
     Test_ExtPen(); // not implemented yet in ROS
     Test_MetaDC();
     Test_Region();
+    Test_CursorIcon();
```
Comment (GeoB99) [inline modules/rostests/apitests/gdi32/GetObject.c:731]: Why calling the test function two times?

**Review by GeoB99 — CHANGES_REQUESTED** [review]

File: modules/rostests/apitests/gdi32/GetObject.c (hunk 2 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/gdi32/GetObject.c:731]: When your new test is run in Windows Server 2003 sp2, the results indicate that the correct response is 64 and NOT 32.

gdi32_GetObject_2K3sp2

It appears that this test is not correct to me. Thanks.

Comment (HBelusca) [reply modules/rostests/apitests/gdi32/GetObject.c:731]: See also https://devblogs.microsoft.com/oldnewthing/20210819-00/?p=105572

Comment (zecarlos1957) [reply modules/rostests/apitests/gdi32/GetObject.c:731]: I deduce here that there is a big difference between XP sp2 and sp3 (sp3 reports 32)

Comment (HBelusca) [reply modules/rostests/apitests/gdi32/GetObject.c:731]: Also, try changing your display settings and the virtual machine software in which you run your given windows version...

Comment (zecarlos1957) [reply modules/rostests/apitests/gdi32/GetObject.c:731]: I deduce here that there is a big difference between XP sp2 and sp3 (sp3 reports 32)

File: modules/rostests/apitests/gdi32/GetObject.c (hunk 1 above)
Comment (zecarlos1957) [reply modules/rostests/apitests/gdi32/GetObject.c:708]: You mean GetObject(...) call?

File: modules/rostests/apitests/gdi32/GetObject.c (hunk 2 above)
Comment (HBelusca) [reply modules/rostests/apitests/gdi32/GetObject.c:731]: Also, please see https://stackoverflow.com/questions/68810645/loadcursor-and-mixed-dpi-multiple-monitors
> When the LoadCursor function is used to load a cursor from a resource, the resulting HCURSOR can be used across different monitors and always appears at the correct size.
> ie: typically:
> on a 96 dpi monitor the 32x32 resource is used,
> on a 192 dpi monitor the 64x64 resource is used.

**Comment by zecarlos1957** [comment]
XPsp3TestApi
RosTestApi

**Comment by zecarlos1957** [comment]
@Doug-Lyons can you please show me this test runing in XPsp2 please.

**Comment by Doug-Lyons** [comment]
@zecarlos1957, Here are the results of your most recent patches.

gdi32_GetObject_2K3sp2_01

File: modules/rostests/apitests/gdi32/GetObject.c [hunk 3]
```diff
@@ -701,6 +722,25 @@ Test_Region(void)
...
+    */
+    GetObjectW(hbmMask, sizeof(bmp), &bmp);
+    ok(bmp.bmWidth == (bmp.bmHeight / 2), "ERR UNICODE CursorIcon RECT got %ldx%ld\n", bmp.bmWidth, bmp.bmHeight);
+    ok(bmp.bmHeight == 64, "ERR UNICODE CursorIcon Height got %ld\n", bmp.bmHeight);
+    DeleteObject(hbmMask);
+ 
+    GetObjectA(hbmMask, sizeof(bmp), &bmp);
+    ok(bmp.bmWidth == (bmp.bmHeight / 2), "ERR ANSI CursorIcon RECT got %ldx%ld\n", bmp.bmWidth, bmp.bmHeight);
+    ok(bmp.bmHeight == 64, "ERR ANSI CursorIcon Height got %ld\n", bmp.bmHeight);
+    DeleteObject(hbmMask);
```
Comment (GeoB99) [inline modules/rostests/apitests/gdi32/GetObject.c:741]: Why there's an ANSI version of this...?

Comment (zecarlos1957) [reply modules/rostests/apitests/gdi32/GetObject.c:741]: You ask for UNICODE version. I just know de UNICODE version that is GetObjectW() and ANSY version that is GetObjectA(). Can you be more specific about your first comment?

Comment (GeoB99) [reply modules/rostests/apitests/gdi32/GetObject.c:741]: You have had a mix of Wide and ANSI calls before I have written the first commit and my comment was that you should only use the Unicode variants of the API calls you use. This is pretty self-explanatory, isn't it?

**Comment by zecarlos1957** [comment]
@binarymaster  This PR seems to me obsolete. Can I delete it?

**Comment by binarymaster** [comment]
@zecarlos1957 obsolete in what sense?

**Review by HBelusca — COMMENTED** [review]
ok I see I had a review in pending state :oof:

File: modules/rostests/apitests/gdi32/GetObject.c [hunk 4]
```diff
@@ -11,6 +11,27 @@
 
 #define INVALID_POINTER ((PVOID)(ULONG_PTR)0xdeadbeefdeadbeefULL)
 
+unsigned char src_mask[]={
```
Comment (HBelusca) [inline modules/rostests/apitests/gdi32/GetObject.c:14]:
```suggestion
UCHAR src_mask[] = {
```

File: modules/rostests/apitests/gdi32/GetObject.c [hunk 5]
```diff
@@ -701,6 +722,20 @@ Test_Region(void)
     DeleteObject(hRgn);
 }
 
+void Test_CursorIcon()
+{
+    BITMAP bmp;
+    HBITMAP hbmMask = CreateBitmap(32, 64, 1, 1, (VOID*)src_mask);
```
Comment (HBelusca) [inline modules/rostests/apitests/gdi32/GetObject.c:728]:
```suggestion
    HBITMAP hbmMask = CreateBitmap(32, 64, 1, 1, src_mask);
```

File: modules/rostests/apitests/gdi32/GetObject.c [hunk 6]
```diff
@@ -701,6 +722,20 @@ Test_Region(void)
...
+    /*
+         This test is valid until Win2003Srv sp2.
```
Comment (HBelusca) [inline modules/rostests/apitests/gdi32/GetObject.c:730]: What happens on Vista+ systems?

Comment (zecarlos1957) [reply modules/rostests/apitests/gdi32/GetObject.c]: Windows7 report 32x64 bitmap. In Vista i dont know

Comment (HBelusca) [reply modules/rostests/apitests/gdi32/GetObject.c]: Can you check whether this depends on the current Windows' display settings? (256 colors, 16-bit color or higher)

Comment (zecarlos1957) [reply modules/rostests/apitests/gdi32/GetObject.c]: RosTest
Here it shows 4 tests in the following order:
800x600x256, 1024x768x16bits, 1600x1200x32bits and 800x600x24bits

Comment (HBelusca) [reply modules/rostests/apitests/gdi32/GetObject.c]: I was asking testing that on Windows 7, not ReactOS.

Comment (zecarlos1957) [reply modules/rostests/apitests/gdi32/GetObject.c]: Win7Test
16 and 32 bits from 800x600, 1280x600 and 1366x768 resolution

**Review by JoachimHenze — CHANGES_REQUESTED** [review]
A lot of screenshots have been posted here, but not yet the most important one: A 2k3sp2 screenshot *showing 0 errors*. Therefore I do request changes.

**Comment by Doug-Lyons** [comment]
W2K3sp2 gdi32_apitest of GetObject with patch attached.

W2K3SP2_gdi32_apitest_GetObject

**Review by JoachimHenze — APPROVED** [review]

**Review by HBelusca — APPROVED** [review]
Should be fine now.

**Comment by zecarlos1957** [comment]
My thanks to @Doug-Lyons

**Pull request merged**